### PR TITLE
Feature/tao 10454/implement search retention

### PR DESCRIPTION
**Pull request by adriloma**
**related to:** https://oat-sa.atlassian.net/browse/TAO-10454

This PR is a continuation of [this one](https://github.com/oat-sa/tao-core-ui-fe/pull/170), and includes:

- Implement search results preservation as detailed on ticket
- Refactor of `searchModal` component to match with other TAO FE components
- Refactor of tests to reach a 100% of coverage

This PR is a dependency of [this](https://github.com/oat-sa/tao-core/pull/2660) PR on `tao-core`, which makes this component accessible for TAO users.

**How to test**
To be able to test this PR locally you need to setup your TAO instance to manage the new Advanced search feature. To do so you need to:
- On your root `composer.json` set the following dependencies:
```
    "oat-sa/tao-core": "feature/TAO-10453/implement-query-builder",
    "oat-sa/generis": "dev-feature/TAO-10203-advanced-search as 12.33.2",
    "oat-sa/extension-tao-dac-simple": "dev-feature/TAO-10203-advanced-search",
    "oat-sa/lib-tao-elasticsearch": "dev-feature/TAO-10203-advanced-search as 1.1.1"
```
- run `composer install`
- if you work with `taoDockerize` run `taoDockerize install`  (database will be reset)
- If you do not user `taoDockerize` run `php tao/scripts/taoUpdate.php`

Now `searchModal` component should be used on search trigger instead of previous search component. Notice that advanced search syntax is not enabled if you do not set up an elastic search server, which is not required to test this feature.

Please let me know if you need more details about how to test the feature.

search-retention

### CODE REVIEW

**Review by andrey-ponamarev — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [ ] The feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request's target is not `master`

File: src/searchModal.js [hunk 1]
```diff
@@ -206,5 +272,12 @@ export default function searchModalFactory(config) {
...
 
-    return instance.init(config);
+    // return initialized instance of searchModal
+    return instance.init(
+        _.defaults(config, {
+            renderTo: 'body',
+            query: '',
+            searchOnInit: true
+        })
+    );
```
Comment (jsconan) [inline src/searchModal.js:282]: The default values are handled by the component factory: `component(api, defaults)`.

**Review by jsconan — COMMENTED** [review]
Code review only. I didn't check the feature in situation, I only ran the unit tests. 

Be careful with the use of promises. Every promise must have a `catch` clause, at least on the top-level consumer. Moreover, pay attention to the inconsistent state it may produce when promises are mixed with synchronous code. It is better to have an emitter/consumer approach.

Be also careful with the exposed API. Namespaces in events should be reserved for the consumers. See in the comments below for more info.

Please also pay attention to code readability. Functions declared after their use, even if not forbidden, are not considered as a good practice.

**Review Checklist**
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

**Review by bziondik — COMMENTED** [review]
- [x]  New code is covered by tests (if applicable)
- [x]  Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x]  New code is respecting code style rules
- [x]  New code is respecting best practices
- [x]  New code is not subject to concurrency issues (if applicable)
- [ ]  Feature is working correctly on my local machine (if applicable)
- [x]  Acceptance criteria are respected
- [x]  Pull request title and description are meaningful

**Comment by bziondik** [comment]
Please, add package.json and package-lock.jsin

File: src/searchModal.js [hunk 2]
```diff
@@ -18,43 +18,51 @@
...
+ * @param {object} config
+ * @param {object} config.renderTo - DOM element where component will be rendered to
+ * @param {string} config.query - search query to be set on component creation
+ * @param {boolean} config.searchOnInit - if init search must be triggered or not (stored results are used instead)
+ * @param {string} config.url - search endpoint to be set on datatable
+ * @returns {searchModal}
  */
 export default function searchModalFactory(config) {
-    const defaults = {
+    config = _.defaults(config, {
```
Comment (bziondik) [inline src/searchModal.js:41]: Please, check https://github.com/oat-sa/tao-core-ui-fe/blob/b4a4fdd3dd8eb797b092977061580250384cc548/src/component.js#L91-L97
you can pass `defaults` in `component()`
and passed configs can be sent to `init()`

**Review by bziondik — APPROVED** [review]
Code review only

**Pull request merged**